src/Snapshotter.cpp

```diff
@@ -123,23 +165,23 @@ void Snapshotter::writeBagFile(const std::string& path, BagCompression compressi
                 if (0 != setpriority(PRIO_PROCESS, 0, 19))
                 {
                     const std::string msg = "setpriority failed: " + std::string(std::strerror(errno));
-                    ROS_ERROR_STREAM(msg);
+                    RCLCPP_ERROR_STREAM(log, msg);
                     ex = std::make_exception_ptr(BagWriteException(msg));
                     return;
                 }
             }
-            bag.open(path, bagmode::Write);
+            writer.open(path);
             /** write all old latched messages 3 seconds before the actual log starts.
              *  The value 3 is arbitrary. The idea is to make the old latched messages stand out
-             *  to a human reader when looking at the bag. We do the calculation in double because ros::Time will throw
-             * when
-             *  the time becomes negative (which can happen when running in simulation because sim time starts at 0) */
-            double latchedTime = bufferCopy->getOldestReceiveTime().toSec() - 3.0;
-            latchedTime = std::max(latchedTime, 0.0);
-            latchedBufferCopy->writeToBag(bag, ros::Time{latchedTime});
-
-            bufferCopy->writeToBag(bag);
-            bag.close();
+             *  to a human reader when looking at the bag. We do the calculation in double because rclcpp::Time will
+             * throw when the time becomes negative (which can happen when running in simulation because sim time starts
+             * at 0) */
+            rclcpp::Time latchedTime = bufferCopy->getOldestReceiveTime() - std::chrono::seconds(3);
+            latchedTime = std::max(latchedTime, rclcpp::Time(static_cast<int64_t>(0), RCL_ROS_TIME));
+            latchedBufferCopy->writeToBag(writer, metaData, latchedTime);
```

Comment:
what happens with multiple latched messages? order can change if multiple are set to 0.

Reply:
then we would have received multiple messages a time zero, on the same topic, which would be a bug anyways.

Reply:
I just checked, the latest sample with the same timestamp is held in the bufferCopy, so this should work as expected.

Reply:
If multiple messages are received in the first 3 seconds, all messages are stored with timestamp 0.